### ISSUE TITLE
AntiCheat is not detecting any integrity changes done on the client

### ISSUE THREAD

**Issue by SeaFood6913** [PlayEveryWare/eos_plugin_for_unity#449]
Hi

I have hard time figuring out what is happening. I think I did things as explained in the Epic documentation.

Here is my setup:

- Client: I have an Anti-cheat client enabled program that I launch with the EAC launcher and on which I passed the anticheat_integritytool that gave back the certificates. The program starts well and it is able to grab the AntiCheatClient interface. The service is also started while the program is launched, so it seems everything is ok. For information, my init code is the following on the client:

```
...
AddNotifyMessageToServerOptions addNotifyMessageToServerOptions = new();
_notificationMessageToServer = _antiCheatInterface.AddNotifyMessageToServer(ref addNotifyMessageToServerOptions, null, OnMessageToServer);

AddNotifyClientIntegrityViolatedOptions addNotifyClientIntegrityViolatedOptions = new();

_notificationClientIntegrityViolated = _antiCheatInterface.AddNotifyClientIntegrityViolated(refaddNotifyClientIntegrityViolatedOptions, null, OnClientIntegrityViolated);

BeginSessionOptions beginSessionOptions = new()
{
  Mode = AntiCheatClientMode.ClientServer,
  LocalUserId = _userId,
};

var result = _antiCheatInterface.BeginSession(ref beginSessionOptions);
if (result != Epic.OnlineServices.Result.Success)
{
  ABLogger.AntiCheat.LogError("Couldn't start session");
  return false;
}

_serverSteamId = default;
Debug.Log("Session started. Sending registration info");

SendRegistrationInfoToAntiCheatServer(serverSteamId);
... 

```

- Server: I launch the AntiCheatserver-enabled program in the Unity Editor. I can see the logs, showing that the session was well started. it matches this code, for information:

```
...
AddNotifyMessageToClientOptions addNotifyMessageToClientOptions = new();

_notificationMessageToClient = _antiCheatInterface.AddNotifyMessageToClient(ref addNotifyMessageToClientOptions, null, OnMessageToClient);

AddNotifyClientActionRequiredOptions addNotifyClientActionRequiredOptions = new();

_notificationClientActionRequired = _antiCheatInterface.AddNotifyClientActionRequired(ref addNotifyClientActionRequiredOptions, null, OnClientActionRequired);

BeginSessionOptions beginSessionOptions = new()
{
  RegisterTimeoutSeconds = 60,
  ServerName = "ReignOfDwarf AntiCheatServer",
  EnableGameplayData = false,
  LocalUserId = null,
};

var result = _antiCheatInterface.BeginSession(ref beginSessionOptions);

if (result != Epic.OnlineServices.Result.Success)
{
  ABLogger.AntiCheat.LogError($"Couldn't start session. Result : {result}");
  return false;
}
_sessionStarted = true;
...

```
Later on, when I start the client I can see on the server logs that the session was well started and that opaque messages are going back and forth (see image)
 
Now, if I modify one of the files that were hashed by the anticheat_integritytool.exe, I expect callbacks to be called, but I get neither the OnClientActionRequired callback on the server, nor the OnClientIntegrityViolated callback on the client...

What could be wrong ?

Attached some log files that do not show any issue

Tx!

SeaFood
![snaplogs](https://github.com/PlayEveryWare/eos_plugin_for_unity/assets/139745960/81ead16e-8e7d-4d71-b4bf-388fed236f31)
[service.log](https://github.com/PlayEveryWare/eos_plugin_for_unity/files/13229468/service.log)
[anticheatlauncher.log](https://github.com/PlayEveryWare/eos_plugin_for_unity/files/13229469/anticheatlauncher.log)

**Comment by paulhazen** [comment]
@SeaFood6913 Apologies - I had written up a response to this issue, but apparently didn't actually submit it yesterday :(

## Preliminary Questions

### Personal Sanity Check
Just checking that this is a typo introduced from copy-pasting the code into GitHub (since this error would cause a compilation error). Anyways - can't hurt to double check, this _is_ a transfer typo, correct?

<img width="400" alt="image" src="https://github.com/PlayEveryWare/eos_plugin_for_unity/assets/1439891/29760c3c-474e-4d4c-bc85-6e86c773f13f">

### Plugin-Specific EAC Documentation
Would I be correct to assume that (in addition to following Epic's documentation on the topic) you followed [this](https://github.com/PlayEveryWare/eos_plugin_for_unity/blob/stable/docs/easy_anticheat_configuration.md) document that describes how to configure EAC in conjunction with the EOS Plugin?

## Suggested Next Steps

I admit that I'm not intimately familiar with the inner workings of EAC (something I'll be rectifying shortly). What I _do_ know about EAC is that the file hash stuff is triggered on subsequent launches of the application. It makes sense to me that there would be a callback for run-time detected changes to game files - but I'm not certain how to trigger that kind of check - I just know that it's _definitely_ triggered when the game is launched.

I would suggest the following steps to further diagnose:
1. Open game with EAC - verify that the file hashes are created.
2. Close game
3. Edit game file(s)
4. Re-launch game, watch for integrity violation notices.

Can you try those steps and get back to us here with the results?

**Comment by SeaFood6913** [comment]
Tx @paulhazen 

Here are some feedbacks:

**Personal Sanity check**
I guess this was a type wile copy pasting, as it did indeed compile properly :)

**Plugin-specific EAC documentation** 
I didn't know this page, thanks for forwarding ! Actually I was applying the integrity tool once the whole project was built and was copying some files from your UPM package. 

I have statements and questions however relating to this page:
- All the prerequisites were already done. Please note however that [EOS documentation](https://dev.epicgames.com/docs/game-services/anti-cheat/anti-cheat-interfaces#client-credentials-and-policy-configuration) states that Client ID should be different for server and client, to not have too much permissions being granted. However, the EOS Plugin "Portal" window has only one entry for Client ID and Client Secret: shouldn't those be differentiated between server and client ?
- What happens if I build the server with "Use EAC" on ? I mean: I assume that if "Is Server" toggle is switched off on the portal menu, then the "Use EAC" should be irrelevant, right ?
- What is the default config for integrity tool if no config file is specified ?
- I am currently building the client with "Use EAC" switched on. It's a pretty long process: I'll let you know the result

**Sugggested Next Steps**
I will check how I can handle what you propose, once the above built is done. However I think we have different understanding here. For me, both the server and the client shall be notified whenever the integrity of one the file initially hashed was modified. This was confirmed by EPIC as I initially posted my request on their forum [as this link](https://eoshelp.epicgames.com/s/question/0D54z00009nkEuhCAE/anticheat-is-not-detecting-any-integrity-changes-done-on-the-client?language=en_US). So I expect the callbacks you saw in my code to be called on such event.

**Comment by SeaFood6913** [comment]
OK I was able to make some further tests, using your above mentionned documentation. Basically, the system behaves like before and I have same issue than the one reported in initial Epic forum.

What I see:
- The certificates were correctely built by integrity tool for the client
- The anti cheat session starts well between client and server, with opaque messages being exchanged

However:
- Nothing occurs, neither on server nor on client whenever I am altering one of the file that was listed to be hashed by integrity tool (the file I chose was a text file)
- An altered file doesn't prevent the client to restart, as you asked...

It seems I am missing something but what ???

**Comment by paulhazen** [comment]
> OK I was able to make some further tests, using your above mentionned documentation. Basically, the system behaves like before and I have same issue than the one reported in initial Epic forum.
> 
> What I see:
> 
> * The certificates were correctely built by integrity tool for the client
> * The anti cheat session starts well between client and server, with opaque messages being exchanged
> 
> However:
> 
> * Nothing occurs, neither on server nor on client whenever I am altering one of the file that was listed to be hashed by integrity tool (the file I chose was a text file)
> * An altered file doesn't prevent the client to restart, as you asked...
> 
> It seems I am missing something but what ???
> 
> One question on the suggested next steps: `Open game with EAC - verify that the file hashes are created.` --> Do you mean the file easyanticheat_wow64_x64.eac that is generated under AppData/Roaming/EasyAntiCheat/ ? If yes, then I do have a 16Mb file that is indeed generated. I also checked that the loader.log loaded correctly (it's ok, see attached) and the EasyAntiCheat_EOS service starts whenever I start the game with the EACLauncher [loader.log](https://github.com/PlayEveryWare/eos_plugin_for_unity/files/13257060/loader.log)

Yes - you correctly understood my suggested next step. Sorry that's not working for you. I'm taking another look at this now and will get back to you with some additional suggestions. I went and read through the question you posted on the epic forum as well, in case there was any additional context doing so might give me. 

> * All the prerequisites were already done. Please note however that [EOS documentation](https://dev.epicgames.com/docs/game-services/anti-cheat/anti-cheat-interfaces#client-credentials-and-policy-configuration) states that Client ID should be different for server and client, to not have too much permissions being granted. However, the EOS Plugin "Portal" window has only one entry for Client ID and Client Secret: shouldn't those be differentiated between server and client ?
> * What happens if I build the server with "Use EAC" on ? I mean: I assume that if "Is Server" toggle is switched off on > * the portal menu, then the "Use EAC" should be irrelevant, right ?
> * What is the default config for integrity tool if no config file is specified ?

...

> Additional topic
> We're using a CI environment for building our client and our server. Would you have a command line equivalent script for applying EAC into the build process?

Tomorrow I will have some answers for the questions you pose in the comment before this one regarding EAC configuration as well as CI support for some of this. I will have to do a little bit of digging myself, but it'll be the first thing I do tomorrow morning, and I apologize again for the delay! :)

**Comment by SeaFood6913** [comment]
Hi Paul

Did you succeed in investigating further ? :)

**Comment by paulhazen** [comment]
> * All the prerequisites were already done. Please note however that [EOS documentation](https://dev.epicgames.com/docs/game-services/anti-cheat/anti-cheat-interfaces#client-credentials-and-policy-configuration) states that Client ID should be different for server and client, to not have too much permissions being granted. However, the EOS Plugin "Portal" window has only one entry for Client ID and Client Secret: shouldn't those be differentiated between server and client?

I'm not seeing where the document you link to above mentions Client ID or Client Secret, can you point that part out for me?

> * What happens if I build the server with "Use EAC" on? I mean: I assume that if "Is Server" toggle is switched off on the portal menu, then the "Use EAC" should be irrelevant, right?

> * What is the default config for integrity tool if no config file is specified?

If none is specified, then the default configuration should be found at `Assets/Plugins/Standalone/Editor/anticheat_integritytool.cfg`

> **Additional topic** We're using a CI environment for building our client and our server. Would you have a command line equivalent script for applying EAC into the build process?

Currently, we don't have such a script.

**Comment by SeaFood6913** [comment]
Hi Paul

`I'm not seeing where the document you link to above mentions Client ID or Client Secret, can you point that part out for me?`
Well exactly at the URL I gave just above :): [EOS Documentation](https://dev.epicgames.com/docs/game-services/anti-cheat/anti-cheat-interfaces#client-credentials-and-policy-configuration)

It states:

`For the Anti-Cheat Client interface, you should use the predefined GameClient policy type. The User Required flag should always be enabled for game clients.

For the Anti-Cheat Server interface and a game using only trusted official servers, you should use the predefined TrustedServer policy type and keep the credential secret.`

You didn't answer this question:
`What happens if I build the server with "Use EAC" on? I mean: I assume that if "Is Server" toggle is switched off on the portal menu, then the "Use EAC" should be irrelevant, right?`

Tx

**Comment by paulhazen** [comment]
The more I look into this problem, the less straightforward I am realizing it is. I suspect I will be better equipped to help you with this after I go play around with EAC for a bit myself. Specifically, I'm going to try and get EAC to trigger a file integrity violation in a sample scene, so I have a better understanding of the mechanics. As for your last question, I apologize for missing it. Yes, it is irrelevant, your assumption is correct.

I will get back to you here in the next few business days at the latest with some more suggestions and some better context for myself.

**Comment by paulhazen** [comment]
Hi @SeaFood6913. I've updated our [documentation](https://github.com/PlayEveryWare/eos_plugin_for_unity/blob/stable/docs/easy_anticheat_configuration.md) describing how to test the EAC integrity tool, can you take a look at it and see if following it helps illuminate what the source of your issue may be?

**Comment by SeaFood6913** [comment]
Hi @paulhazen , 

Tx for the update. Actually everthing works as you described. However I was expecting slightly different behaviour:
(Please note that our topology is Server- client, but a client can connect to the server of this choice - within a list displayed to him)
- While your scenario is triggering a client intergrity event on the client, it does't do anything neither on the server nor on the Epic web interface, which sounds odd to me
- When running the client, if I attempt to change a hashed file, I was expecting the modification to be immediately detected and reported. I happen not to be able to modify at all such file whenever the client is active

**Comment by paulhazen** [comment]
> Tx for the update. Actually everthing works as you described. However I was expecting slightly different behaviour: (Please note that our topology is Server- client, but a client can connect to the server of this choice - within a list displayed to him)

That's good to hear!

> * While your scenario is triggering a client intergrity event on the client, it does't do anything neither on the server nor on the Epic web interface, which sounds odd to me.

That is interesting. I suppose it's not a component of Anti-Cheat that Epic considered to be a vulnerable enough thread-factor to warrant protecting. You could always implement something of your own - but I suppose that was the whole point of using EAC - not having to make a bespoke solution.

> * When running the client, if I attempt to change a hashed file, I was expecting the modification to be immediately detected and reported. I happen not to be able to modify at all such file whenever the client is active.

This was my initial assumption as well, but like you I was unable to change the file with the client active. The more I thought about it the more I became comfortable that by checking the files upon each launch, and by making the files unalterable during runtime, things were pretty darn secure. Seems like such a solution is as secure as the operating system that's running it. That is, if an attacker could alter the file during runtime, because it wouldn't trigger a call, the file could be re-read by the game and cheating could perhaps happen. I think this is mitigated by the fact that the EAC bootstrapper ensures that those files are as locked as the operating system is capable of locking them, effectively closing that loophole. If the files can still be altered, that's a failure of the OS' security, not the application that correctly implemented it's security features.

I'm going to go ahead and close this issue - I really appreciate your continued dialogue, this was a great opportunity for me to learn more about this area. If you feel like the issue needs re-opening, feel free to do so with any additional context you may have. Thanks for using the plug-in and helping us improve our documentation! :)

**Comment by SeaFood6913** [comment]
Tx @paulhazen, for keeping investigating ;) ! Happy it helped for your doc !
